File: internal/app/macos_code_identity.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"os/exec"
	"path/filepath"
	"runtime"
	"strings"
)

const (
	macOSCodeIdentityAdhoc            = "adhoc"
	macOSCodeIdentityDeveloperID      = "developer-id"
	macOSCodeIdentityInvalid          = "invalid"
	macOSCodeIdentityLocalCertificate = "local-certificate"
	macOSCodeIdentityNotApplicable    = "not-applicable"
	macOSCodeIdentityUnknown          = "unknown"
	macOSCodeIdentityUnsigned         = "unsigned"
)

const (
	macOSPermissionStabilityInvalid             = "invalid"
	macOSPermissionStabilityLocalMachineStable  = "local-machine-stable"
	macOSPermissionStabilityNotApplicable       = "not-applicable"
	macOSPermissionStabilityPublicStable        = "public-stable"
	macOSPermissionStabilityUnstableAfterUpdate = "unstable-after-update"
)

type macOSCodeIdentity struct {
	BundleID                            string
	CodeIdentifier                      string
	Signature                           string
	CDHash                              string
	TeamIdentifier                      string
	Authorities                         []string
	DesignatedRequirement               string
	DesignatedRequirementIsCDHashOnly   bool
	IdentityKind                        string
	PermissionStability                 string
	StableRequirementFingerprint        string
	CodesignVerifyOutput                string
	CodesignDisplayOutput               string
	CodesignDesignatedRequirementOutput string
}

type macOSCodeIdentityInspector func(string) (macOSCodeIdentity, error)

func inspectMacOSAppCodeIdentity(appPath string) (macOSCodeIdentity, error) {
	if runtime.GOOS != "darwin" {
		return macOSCodeIdentity{
			IdentityKind:        macOSCodeIdentityNotApplicable,
			PermissionStability: macOSPermissionStabilityNotApplicable,
		}, nil
	}

	appPath = strings.TrimSpace(appPath)
	if appPath == "" {
		return macOSCodeIdentity{}, fmt.Errorf("app bundle path is required")
	}

	bundleID := readPlistRaw(filepath.Join(appPath, "Contents", "Info.plist"), "CFBundleIdentifier")
	verifyOutput, verifyErr := exec.Command("/usr/bin/codesign", "--verify", "--deep", "--strict", "--verbose=2", appPath).CombinedOutput()
	displayOutput, displayErr := exec.Command("/usr/bin/codesign", "-dv", appPath).CombinedOutput()
	requirementOutput, requirementErr := exec.Command("/usr/bin/codesign", "-d", "-r-", appPath).CombinedOutput()

	identity := parseMacOSCodeIdentity(
		bundleID,
		verifyErr == nil,
		string(verifyOutput),
		string(displayOutput),
		string(requirementOutput),
	)
	if verifyErr != nil {
		return identity, fmt.Errorf("%w: %s", verifyErr, strings.TrimSpace(string(verifyOutput)))
	}
	if displayErr != nil {
		return identity, fmt.Errorf("codesign display failed: %w: %s", displayErr, strings.TrimSpace(string(displayOutput)))
	}
	if requirementErr != nil {
		return identity, fmt.Errorf("codesign designated requirement failed: %w: %s", requirementErr, strings.TrimSpace(string(requirementOutput)))
	}
	return identity, nil
}

func parseMacOSCodeIdentity(bundleID string, verified bool, verifyOutput string, displayOutput string, requirementOutput string) macOSCodeIdentity {
	identity := macOSCodeIdentity{
		BundleID:                            strings.TrimSpace(bundleID),
		CodeIdentifier:                      codesignLineValue(displayOutput, "Identifier"),
		Signature:                           codesignLineValue(displayOutput, "Signature"),
		CDHash:                              codesignLineValue(displayOutput, "CDHash"),
		TeamIdentifier:                      normalizeCodesignUnsetValue(codesignLineValue(displayOutput, "TeamIdentifier")),
		Authorities:                         codesignLineValues(displayOutput, "Authority"),
		DesignatedRequirement:               extractDesignatedRequirement(requirementOutput),
		CodesignVerifyOutput:                strings.TrimSpace(verifyOutput),
		CodesignDisplayOutput:               strings.TrimSpace(displayOutput),
		CodesignDesignatedRequirementOutput: strings.TrimSpace(requirementOutput),
	}
	if identity.BundleID == "" {
		identity.BundleID = identity.CodeIdentifier
	}
	identity.DesignatedRequirementIsCDHashOnly = isCDHashOnlyDesignatedRequirement(identity.DesignatedRequirement)
	identity.IdentityKind = inferMacOSCodeIdentityKind(identity, verified)
	identity.PermissionStability = permissionStabilityForMacOSCodeIdentity(identity.IdentityKind)
	if identity.DesignatedRequirement != "" {
		identity.StableRequirementFingerprint = fingerprintMacOSDesignatedRequirement(identity.DesignatedRequirement)
	}
	return identity
}

func codesignLineValue(output string, key string) string {
	values := codesignLineValues(output, key)
	if len(values) == 0 {
		return ""
	}
	return values[0]
}

func codesignLineValues(output string, key string) []string {
	prefix := key + "="
	var values []string
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, prefix) {
			values = append(values, strings.TrimSpace(strings.TrimPrefix(line, prefix)))
		}
	}
	return values
}

func normalizeCodesignUnsetValue(value string) string {
	value = strings.TrimSpace(value)
	if strings.EqualFold(value, "not set") {
		return ""
	}
	return value
}

func extractDesignatedRequirement(output string) string {
	for _, line := range strings.Split(output, "\n") {
		line = strings.TrimSpace(line)
		if strings.Contains(line, "designated =>") {
			return line
		}
	}
	return strings.TrimSpace(output)
}

func isCDHashOnlyDesignatedRequirement(requirement string) bool {
	requirement = strings.TrimSpace(requirement)
	requirement = strings.TrimPrefix(requirement, "#")
	requirement = strings.TrimSpace(requirement)
	return strings.HasPrefix(requirement, "designated => cdhash ")
}

func inferMacOSCodeIdentityKind(identity macOSCodeIdentity, verified bool) string {
	if !verified {
		if strings.Contains(identity.CodesignVerifyOutput, "not signed") ||
			strings.Contains(identity.CodesignDisplayOutput, "not signed") ||
			strings.Contains(identity.CodesignDesignatedRequirementOutput, "not signed") {
			return macOSCodeIdentityUnsigned
		}
		return macOSCodeIdentityInvalid
	}
	if strings.EqualFold(identity.Signature, "adhoc") || identity.DesignatedRequirementIsCDHashOnly {
		return macOSCodeIdentityAdhoc
	}
	for _, authority := range identity.Authorities {
		if strings.Contains(authority, "Developer ID Application:") {
			return macOSCodeIdentityDeveloperID
		}
	}
	if len(identity.Authorities) > 0 {
		return macOSCodeIdentityLocalCertificate
	}
	if identity.Signature == "" {
		return macOSCodeIdentityUnsigned
	}
	return macOSCodeIdentityUnknown
}

func permissionStabilityForMacOSCodeIdentity(identityKind string) string {
	switch identityKind {
	case macOSCodeIdentityDeveloperID:
		return macOSPermissionStabilityPublicStable
	case macOSCodeIdentityLocalCertificate:
		return macOSPermissionStabilityLocalMachineStable
	case macOSCodeIdentityNotApplicable:
		return macOSPermissionStabilityNotApplicable
	case macOSCodeIdentityInvalid:
		return macOSPermissionStabilityInvalid
	default:
		return macOSPermissionStabilityUnstableAfterUpdate
	}
}

func isPermissionStableMacOSCodeIdentity(identity macOSCodeIdentity) bool {
	if identity.DesignatedRequirementIsCDHashOnly || identity.StableRequirementFingerprint == "" {
		return false
	}
	return identity.IdentityKind == macOSCodeIdentityDeveloperID ||
		identity.IdentityKind == macOSCodeIdentityLocalCertificate
}

func fingerprintMacOSDesignatedRequirement(requirement string) string {
	requirement = strings.TrimSpace(requirement)
	if requirement == "" {
		return ""
	}
	sum := sha256.Sum256([]byte(requirement))
	return hex.EncodeToString(sum[:])
}

func verifyPermissionStableMacOSUpdateCandidate(appPath string, inspect macOSCodeIdentityInspector) (macOSCodeIdentity, error) {
	if inspect == nil {
		return macOSCodeIdentity{}, nil
	}
	identity, err := inspect(appPath)
	if err != nil {
		return identity, err
	}
	if identity.IdentityKind == macOSCodeIdentityNotApplicable {
		return identity, nil
	}
	if !isPermissionStableMacOSCodeIdentity(identity) {
		return identity, fmt.Errorf("identityKind=%s permissionStability=%s cdhashOnly=%t",
			identity.IdentityKind,
			identity.PermissionStability,
			identity.DesignatedRequirementIsCDHashOnly,
		)
	}
	return identity, nil
}

func validateAutoUpdateCodeIdentityTransition(currentAppPath string, stagedAppPath string, inspect macOSCodeIdentityInspector) (macOSCodeIdentity, string, error) {
	if inspect == nil {
		return macOSCodeIdentity{}, "", nil
	}
	currentIdentity, err := inspect(currentAppPath)
	if err != nil {
		return macOSCodeIdentity{}, "", fmt.Errorf("current Arlecchino.app macOS code identity could not be inspected: %w", err)
	}
	stagedIdentity, err := verifyPermissionStableMacOSUpdateCandidate(stagedAppPath, inspect)
	if err != nil {
		return stagedIdentity, "", fmt.Errorf("staged Arlecchino.app macOS code identity is not permission-stable: %w", err)
	}
	if stagedIdentity.IdentityKind == macOSCodeIdentityNotApplicable {
		return stagedIdentity, "", nil
	}
	if currentIdentity.BundleID == "" || stagedIdentity.BundleID == "" || currentIdentity.BundleID != stagedIdentity.BundleID {
		return stagedIdentity, "", fmt.Errorf("staged Arlecchino.app bundle id %q does not match current app bundle id %q", stagedIdentity.BundleID, currentIdentity.BundleID)
	}

	if isPermissionStableMacOSCodeIdentity(currentIdentity) {
		if currentIdentity.StableRequirementFingerprint == "" ||
			currentIdentity.StableRequirementFingerprint != stagedIdentity.StableRequirementFingerprint {
			return stagedIdentity, "", fmt.Errorf("staged Arlecchino.app uses a different macOS signing identity; refusing update to avoid resetting folder permissions")
		}
		return stagedIdentity, "", nil
	}

	if currentIdentity.IdentityKind == macOSCodeIdentityAdhoc && isPermissionStableMacOSCodeIdentity(stagedIdentity) {
		return stagedIdentity, "This update migrates Arlecchino to a stable macOS signing identity. macOS may ask for folder access one final time after relaunch.", nil
	}

	return stagedIdentity, "", fmt.Errorf("current Arlecchino.app identityKind=%s cannot be safely migrated by automatic update", currentIdentity.IdentityKind)
}

File: scripts/macos-code-identity.mjs
#!/usr/bin/env node

import crypto from "node:crypto";
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawnSync } from "node:child_process";
import { pathToFileURL } from "node:url";

export const identityKinds = {
  adhoc: "adhoc",
  developerID: "developer-id",
  invalid: "invalid",
  localCertificate: "local-certificate",
  unsigned: "unsigned",
  unknown: "unknown",
};

export const permissionStabilities = {
  invalid: "invalid",
  localMachineStable: "local-machine-stable",
  publicStable: "public-stable",
  unstableAfterUpdate: "unstable-after-update",
};

const runCommand = (command, args) => {
  const result = spawnSync(command, args, { encoding: "utf8" });
  return {
    status: result.status ?? 1,
    stdout: result.stdout || "",
    stderr: result.stderr || "",
    output: `${result.stdout || ""}${result.stderr || ""}`.trim(),
  };
};

export const parseCodesignDisplay = (output) => {
  const lineValues = (key) =>
    output
      .split(/\r?\n/)
      .map((line) => line.trim())
      .filter((line) => line.startsWith(`${key}=`))
      .map((line) => line.slice(key.length + 1).trim());
  const first = (key) => lineValues(key)[0] || "";
  const teamIdentifier = first("TeamIdentifier");
  return {
    codeIdentifier: first("Identifier"),
    signature: first("Signature"),
    cdHash: first("CDHash"),
    teamIdentifier:
      teamIdentifier.toLowerCase() === "not set" ? "" : teamIdentifier,
    authorities: lineValues("Authority"),
    raw: output.trim(),
  };
};

export const extractDesignatedRequirement = (output) => {
  const line = output
    .split(/\r?\n/)
    .map((item) => item.trim())
    .find((item) => item.includes("designated =>"));
  return line || output.trim();
};

export const isCDHashOnlyDesignatedRequirement = (requirement) => {
  const normalized = requirement.trim().replace(/^#\s*/, "");
  return normalized.startsWith("designated => cdhash ");
};

export const permissionStabilityForIdentityKind = (identityKind) => {
  switch (identityKind) {
    case identityKinds.developerID:
      return permissionStabilities.publicStable;
    case identityKinds.localCertificate:
      return permissionStabilities.localMachineStable;
    case identityKinds.invalid:
      return permissionStabilities.invalid;
    default:
      return permissionStabilities.unstableAfterUpdate;
  }
};

export const inferIdentityKind = ({
  verified,
  verifyOutput,
  display,
  designatedRequirement,
}) => {
  if (!verified) {
    if (
      /not signed/i.test(verifyOutput) ||
      /not signed/i.test(display.raw || "") ||
      /not signed/i.test(designatedRequirement)
    ) {
      return identityKinds.unsigned;
    }
    return identityKinds.invalid;
  }
  if (
    display.signature.toLowerCase() === "adhoc" ||
    isCDHashOnlyDesignatedRequirement(designatedRequirement)
  ) {
    return identityKinds.adhoc;
  }
  if (
    display.authorities.some((authority) =>
      authority.includes("Developer ID Application:"),
    )
  ) {
    return identityKinds.developerID;
  }
  if (display.authorities.length > 0) {
    return identityKinds.localCertificate;
  }
  return display.signature ? identityKinds.unknown : identityKinds.unsigned;
};

export const stableRequirementFingerprint = (designatedRequirement) => {
  const normalized = designatedRequirement.trim();
  if (!normalized) return "";
  return crypto.createHash("sha256").update(normalized).digest("hex");
};

export const isPermissionStableIdentity = (identity) =>
  !identity.designatedRequirementIsCdhashOnly &&
  Boolean(identity.stableRequirementFingerprint) &&
  (identity.identityKind === identityKinds.developerID ||
    identity.identityKind === identityKinds.localCertificate);

export const inspectAppBundleCodeIdentity = (appBundle) => {
  const verify = runCommand("/usr/bin/codesign", [
    "--verify",
    "--deep",
    "--strict",
    "--verbose=2",
    appBundle,
  ]);
  const displayResult = runCommand("/usr/bin/codesign", ["-dv", appBundle]);
  const requirementResult = runCommand("/usr/bin/codesign", [
    "-d",
    "-r-",
    appBundle,
  ]);
  const display = parseCodesignDisplay(displayResult.output);
  const designatedRequirement = extractDesignatedRequirement(
    requirementResult.output,
  );
  const identityKind = inferIdentityKind({
    verified: verify.status === 0,
    verifyOutput: verify.output,
    display,
    designatedRequirement,
  });
  const identity = {
    bundleId: readBundleIdentifier(appBundle) || display.codeIdentifier,
    codeIdentifier: display.codeIdentifier,
    signature: display.signature,
    cdHash: display.cdHash,
    teamIdentifier: display.teamIdentifier,
    authorities: display.authorities,
    identityKind,
    permissionStability: permissionStabilityForIdentityKind(identityKind),
    designatedRequirement,
    designatedRequirementIsCdhashOnly: isCDHashOnlyDesignatedRequirement(
      designatedRequirement,
    ),
    stableRequirementFingerprint: stableRequirementFingerprint(
      designatedRequirement,
    ),
    codesignVerifyPassed: verify.status === 0,
  };
  if (
    verify.status !== 0 ||
    displayResult.status !== 0 ||
    requirementResult.status !== 0
  ) {
    const reason = [
      verify.output,
      displayResult.output,
      requirementResult.output,
    ]
      .filter(Boolean)
      .join("\n");
    const error = new Error(
      `codesign identity inspection failed for ${appBundle}: ${reason}`,
    );
    error.identity = identity;
    throw error;
  }
  return identity;
};

export const inspectZippedAppCodeIdentity = (
  zipPath,
  appName = "Arlecchino.app",
) => {
  if (process.platform !== "darwin") {
    throw new Error("macOS updater ZIP code identity checks require macOS");
  }
  const tempDir = fs.mkdtempSync(
    path.join(os.tmpdir(), "arlecchino-update-identity-"),
  );
  try {
    const extract = runCommand("/usr/bin/ditto", [
      "-x",
      "-k",
      zipPath,
      tempDir,
    ]);
    if (extract.status !== 0) {
      throw new Error(`could not extract updater ZIP: ${extract.output}`);
    }
    const appBundle = findAppBundle(tempDir, appName);
    if (!appBundle) {
      throw new Error(`updater ZIP does not contain ${appName}`);
    }
    return inspectAppBundleCodeIdentity(appBundle);
  } finally {
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
};

export const manifestSafeCodeIdentity = (identity) => ({
  bundleId: identity.bundleId,
  codeIdentifier: identity.codeIdentifier,
  signature: identity.signature,
  cdHash: identity.cdHash,
  teamIdentifier: identity.teamIdentifier,
  authorities: identity.authorities,
  identityKind: identity.identityKind,
  permissionStability: identity.permissionStability,
  designatedRequirementIsCdhashOnly: identity.designatedRequirementIsCdhashOnly,
  stableRequirementFingerprint: identity.stableRequirementFingerprint,
});

const readBundleIdentifier = (appBundle) => {
  const infoPlist = path.join(appBundle, "Contents", "Info.plist");
  const result = runCommand("/usr/libexec/PlistBuddy", [
    "-c",
    "Print :CFBundleIdentifier",
    infoPlist,
  ]);
  return result.status === 0 ? result.stdout.trim() : "";
};

const findAppBundle = (root, appName) => {
  const direct = path.join(root, appName);
  if (fs.existsSync(direct) && fs.statSync(direct).isDirectory()) {
    return direct;
  }
  const entries = fs.readdirSync(root, { withFileTypes: true });
  for (const entry of entries) {
    if (!entry.isDirectory() || entry.name === "__MACOSX") continue;
    const candidate = path.join(root, entry.name);
    if (entry.name === appName) {
      return candidate;
    }
    const nested = findAppBundle(candidate, appName);
    if (nested) return nested;
  }
  return "";
};

const invokedAsMain = process.argv[1]
  ? import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href
  : false;

if (invokedAsMain) {
  const target = process.argv[2] || "";
  if (!target) {
    console.error(
      "Usage: scripts/macos-code-identity.mjs <Arlecchino.app|zip>",
    );
    process.exit(2);
  }
  const identity = target.endsWith(".zip")
    ? inspectZippedAppCodeIdentity(target)
    : inspectAppBundleCodeIdentity(target);
  process.stdout.write(`${JSON.stringify(identity, null, 2)}\n`);
}
